cxx/distributions/dirichlet_categorical.hh

```diff
@@ -3,57 +3,81 @@
-    counts[size_t(x)] += 1;
-    ++n;
-  }
-  void unincorporate(const double& x) {
-    const size_t y = x;
-    assert(y < counts.size());
-    counts[y] -= 1;
-    --n;
-    assert(0 <= counts[y]);
-    assert(0 <= n);
-  }
-  double logp(const double& x) const {
-    assert(x >= 0 && x < counts.size());
-    const double numer = log(alpha + counts[size_t(x)]);
-    const double denom = log(n + alpha * counts.size());
-    return numer - denom;
-  }
-  double logp_score() const {
-    const size_t k = counts.size();
-    const double a = alpha * k;
-    const double lg = std::transform_reduce(
-        counts.cbegin(), counts.cend(), 0, std::plus{},
-        [&](size_t y) -> double { return lgamma(y + alpha); });
-    return lgamma(a) - lgamma(a + n) + lg - k * lgamma(alpha);
-  }
-  double sample() {
-    std::vector<double> weights(counts.size());
-    std::transform(counts.begin(), counts.end(), weights.begin(),
-                   [&](size_t y) -> double { return y + alpha; });
-    int idx = choice(weights, prng);
-    return double(idx);
-  }
+    // DirichletCategorical does not take ownership of prng.
+    DirichletCategorical(std::mt19937 *prng, int k) {  // k is number of categories
+        this->prng = prng;
+        counts = std::vector<int>(k, 0);
+        n = 0;
+    }
+    void incorporate(const double& x) {
+        assert(x >= 0 && x < counts.size());
+        counts[size_t(x)] += 1;
+        ++n;
+    }
+    void unincorporate(const double& x) {
+        const size_t y = x;
+        assert(y < counts.size());
+        counts[y] -= 1;
+        --n;
+        assert(0 <= counts[y]);
+        assert(0 <= n);
+    }
+    double logp(const double& x) const {
+        assert(x >= 0 && x < counts.size());
+        const double numer = log(alpha + counts[size_t(x)]);
+        const double denom = log(n + alpha * counts.size());
+        return numer - denom;
+    }
+    double logp_score() const {
+        const size_t k = counts.size();
+        const double a = alpha * k;
+        const double lg = std::transform_reduce(
+            counts.cbegin(),
+            counts.cend(),
+            0,
+            std::plus{},
+            [&](size_t y) -> double {return lgamma(y + alpha); }
+        );
+        return lgamma(a) - lgamma(a + n) + lg - k * lgamma(alpha);
+    }
+    double sample() {
+        std::vector<double> weights(counts.size());
+        std::transform(
+            counts.begin(),
+            counts.end(),
+            weights.begin(),
+            [&](size_t y) -> double { return y + alpha; }
+        );
+        int idx = choice(weights, prng);
+        return double(idx);
+    }
+    void transition_hyperparameters() {
+      std::vector<double> logps;
+      std::vector<double> alphas;
+      // C++ doesn't yet allow range for-loops over existing variables.  Sigh.
+      for (double alphat : ALPHA_GRID) {
+          alpha = alphat;
+          double lp = logp_score();
+          if (!std::isnan(lp)) {
```

Comment:
Do you have a sense of when nans are showing up?

Reply:
None in dirichlet_categorical, at least with the current unit test, but I did see them previously in BetaBernoulli's transition_hyperparamters.

Huh.  When I rerun that now, I no longer see any Nan's.  I wonder if the above lbeta change fixed them.